src/main/java/org/cactoos/io/Zip.java

```diff
@@ -33,50 +33,58 @@
 import java.util.zip.ZipEntry;
 import java.util.zip.ZipOutputStream;
 import org.cactoos.Input;
+import org.cactoos.func.ForEach;
+import org.cactoos.scalar.ItemAt;
+import org.cactoos.text.Joined;
 
 /**
  * Zip files and directory.
  * <br>
  * <br>There is no thread-safety guarantee.
  *
  * @since 0.29
+ * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
  */
 @SuppressWarnings("PMD.AvoidInstantiatingObjectsInLoops")
 public final class Zip implements Input {
 
     /**
-     * Origin directory.
+     * Origin list of files.
      */
-    private final Directory origin;
+    private final Directory directory;
 
     /**
      * Ctor.
      *
-     * @param origin Origin directory.
+     * @param dir Origin directory.
      */
-    public Zip(final Directory origin) {
-        this.origin = origin;
+    public Zip(final Directory dir) {
+        this.directory = dir;
     }
 
     @Override
     public InputStream stream() throws Exception {
         final ByteArrayOutputStream out = new ByteArrayOutputStream();
+        final Path parent = new ItemAt<>(0, this.directory).value().getParent();
         try (ZipOutputStream zip = new ZipOutputStream(out)) {
-            for (final Path path : this.origin) {
-                final File file = path.toFile();
-                final ZipEntry entry = new ZipEntry(
-                    file.getPath()
-                );
-                zip.putNextEntry(entry);
-                if (file.isFile()) {
-                    try (
-                        FileInputStream input = new FileInputStream(file)
-                    ) {
-                        zip.write(new BytesOf(new InputOf(input)).asBytes());
+            new ForEach<Path>(
+                path -> {
+                    final File file = path.toFile();
+                    final String relative = parent.relativize(path).toString();
+                    if (file.isFile()) {
+                        zip.putNextEntry(new ZipEntry(relative));
+                        try (FileInputStream fis = new FileInputStream(file)) {
+                            zip.write(new BytesOf(new InputOf(fis)).asBytes());
+                        }
+                    } else if (!new Directory(file).iterator().hasNext()) {
```

Comment:
@fanifieiev what's happening here? it seems this does not go recursively into directories? Can you add a test to validate it works with sub directories?

Reply:
@fanifieiev you forgot that

Reply:
@victornoel it is outdated, there is a new code instead of that

Reply:
@fanifieiev my comment is still valid, maybe I am not clear? I doubt when reading the code that it does recursively zip the directories, so instead of thinking too much, I'm asking you for a test that validates this expected behaviour.

Reply:
@victornoel It does not do recursively. The `Directory` object provides the iterable of all the available paths.
The test contains subfolders.
So, the test has the following files and directories:
```
/abc/A.txt # file inside root folder
/abc/B/B.txt # file inside subfolder
/abc/C # empty subdirectory
```

Reply:
@fanifieiev I missed that, ok. I think it would make sense to have multiple tests for the various aspects being tested and not everything in the same test though:
- empty dir
- multiple nested directories
- nominal behaviour with the correct base path being zipped

Reply:
@victornoel Agree, but I think we should have at least one test which would have complex case, when there is an empty folder, file in subfolder and just a file inside main folder.
I will also create a separate test for empty dir.